File: services/bff/src/telemetry.ts
// Four-signal OpenTelemetry bootstrap for the BFF service (MELT).
//
// Loaded via `node -r ./dist/telemetry.js` BEFORE the app, so the http server
// (incoming, via @hono/node-server) and grpc clients (outgoing → catalogue/cart)
// are auto-instrumented. Takes the BFF from traces-only to the full four
// correlated signals — Metrics, Events, Logs, Traces — over OTLP/gRPC, matching
// the catalogue (Go) reference and the users (Python) / cart (Node) retrofits
// (ARCHITECTURE.md §9 / ADR-0002). One NodeSDK wires all three providers off its
// auto-detected resource, so every record agrees on service.name /
// service.namespace / deployment.environment (from the OTEL_* env in compose),
// and in-request logs/events carry the active trace_id (the correlation contract).
//
// BFF vs Cart (same OTel-JS family, but an HTTP server not a gRPC server):
//   * RED is HTTP-SERVER-side. @opentelemetry/instrumentation-http (0.218) already
//     emits an HTTP server duration histogram, so — unlike the Go/Python/Cart
//     gRPC interceptor — RED needs no hand-rolled recorder. We opt into the STABLE
//     HTTP semconv (below) so it emits the seconds-scale `http.server.request.duration`
//     with stable attributes (http.request.method, http.response.status_code); a
//     View aligns its buckets with the cross-stack reference. Verified in Docker:
//     the http SERVER span is active inside the Hono fetch handler, so logs/events
//     emitted in handlers carry trace_id/span_id (Hono is fetch-based — confirmed).
//   * METRIC EXEMPLARS ARE NOT AVAILABLE in OpenTelemetry-JS (sdk-metrics 2.7.1,
//     the latest) — the histogram aggregator never records exemplars and the OTLP
//     serializer never writes the exemplar field (settled & documented for Cart;
//     §9 "Per-stack exemplar policy"). The metric→trace link for this service is
//     therefore provided by the LGTM bundle's Tempo metrics-generator
//     (`send_exemplars: true`), which derives exemplar-bearing span metrics
//     (traces_spanmetrics_*{service="bff"}) from the BFF's own traces server-side.
//   * Logs + Events are mature and identical to Cart: the Logs API captures the
//     active trace_id/span_id automatically, and a domain event is a log record
//     with the `eventName` field set.

// Opt into the STABLE HTTP semantic conventions BEFORE the instrumentations are
// constructed (getNodeAutoInstrumentations runs in `new NodeSDK` below, and the
// HttpInstrumentation reads this env in its constructor). This switches the http
// server metric from the legacy ms-scale `http.server.duration` to the stable
// seconds-scale `http.server.request.duration` (verified in Docker).
process.env.OTEL_SEMCONV_STABILITY_OPT_IN ??= 'http';

import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-grpc';
import { PeriodicExportingMetricReader, AggregationType } from '@opentelemetry/sdk-metrics';
import { BatchLogRecordProcessor } from '@opentelemetry/sdk-logs';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { type Attributes } from '@opentelemetry/api';
import { logs, SeverityNumber } from '@opentelemetry/api-logs';

// Instrumentation scope shared by the BFF's own logger and events — distinct from
// the http/grpc auto-instrumentation scopes. Lets queries pick out the BFF
// service's own domain events (in Loki, events carry scope_name="shoeshop/bff").
const SCOPE = 'shoeshop/bff';

// Second-scale buckets for the HTTP server latency histogram. Mirrors the
// Go/Python/Cart reference buckets so the histogram is comparable across stacks.
const LATENCY_BUCKETS = [0.001, 0.0025, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5];

const sdk = new NodeSDK({
  // ── Traces ── unchanged from the existing wiring: one SERVER span per HTTP
  // request, child gRPC CLIENT spans (bff → catalogue, bff → cart → redis).
  traceExporter: new OTLPTraceExporter(),
  // ── Metrics ── periodic OTLP export. RED is the http instrumentation's own
  // `http.server.request.duration` (rate via _count, errors via status_code); the
  // View gives it the cross-stack second-scale buckets.
  metricReaders: [new PeriodicExportingMetricReader({ exporter: new OTLPMetricExporter() })],
  views: [
    {
      instrumentName: 'http.server.request.duration',
      aggregation: {
        type: AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
        options: { boundaries: LATENCY_BUCKETS },
      },
    },
  ],
  // ── Logs + Events ── both flow through one LoggerProvider → OTLP → Loki.
  logRecordProcessors: [new BatchLogRecordProcessor(new OTLPLogExporter())],
  instrumentations: [
    getNodeAutoInstrumentations({
      // fs spans are noisy and unhelpful here.
      '@opentelemetry/instrumentation-fs': { enabled: false },
      // keep the liveness probe out of traces and the RED metric.
      '@opentelemetry/instrumentation-http': {
        ignoreIncomingRequestHook: (req) => (req.url ?? '') === '/healthz',
      },
    }),
  ],
});

sdk.start();

// The logger is created AFTER start(): only then has NodeSDK installed the global
// logger provider (before start, logs.getLogger is a no-op).
const otelLogger = logs.getLogger(SCOPE);

/**
 * Structured logger that fans out to BOTH stdout JSON (so `task logs -- bff` still
 * works) AND OTLP logs → Loki. When called inside a request's span the Logs API
 * stamps the active trace_id/span_id — the trace↔log join key from the correlation
 * contract. Startup logs (outside any span) carry no trace_id, as expected.
 */
function emit(severityNumber: SeverityNumber, level: string, msg: string, attributes?: Attributes): void {
  console.log(JSON.stringify({ level, msg, ...attributes }));
  otelLogger.emit({ severityNumber, severityText: level.toUpperCase(), body: msg, attributes });
}

export const log = {
  info: (msg: string, attributes?: Attributes) => emit(SeverityNumber.INFO, 'info', msg, attributes),
  warn: (msg: string, attributes?: Attributes) => emit(SeverityNumber.WARN, 'warn', msg, attributes),
  error: (msg: string, attributes?: Attributes) => emit(SeverityNumber.ERROR, 'error', msg, attributes),
};

/**
 * Emit a domain/lifecycle event via the Logs API — a log record carrying an
 * eventName (e.g. "bff.product.viewed"). Emitting within a request's span makes
 * the SDK stamp the active trace_id/span_id, so events join to their trace exactly
 * like logs do (the four-signal "E"). The BFF is a read-path aggregator, so its
 * events are modest and meaningful. In Loki, events are told apart from logs by
 * scope_name="shoeshop/bff" + the eventName body. Richer domain events arrive with
 * the v0.3 NATS write path.
 */
export function event(name: string, attributes?: Attributes): void {
  otelLogger.emit({
    eventName: name,
    severityNumber: SeverityNumber.INFO,
    body: name,
    attributes,
  });
}

function shutdown(): void {
  sdk
    .shutdown()
    .then(() => process.exit(0))
    .catch(() => process.exit(1));
}

process.on('SIGTERM', shutdown);
process.on('SIGINT', shutdown);

File: services/payment/src/telemetry.rs
//! Four-signal MELT bootstrap (ADR-0002 §9). One shared `Resource` feeds the
//! tracer, meter and logger providers, all exporting OTLP/gRPC to the LGTM bundle,
//! so every signal agrees on service.name / service.namespace / deployment.environment.
//!
//! Design mirrors the Go/Java references: saga spans are created with the OTel API
//! directly (see `natstrace`), and logs/events go through `tracing` → the
//! opentelemetry-appender-tracing bridge, which stamps each record with the
//! trace_id/span_id of the *attached* OTel context. The resource is built from
//! OTEL_SERVICE_NAME + OTEL_RESOURCE_ATTRIBUTES only (no host.*/process.* detectors),
//! preserving the cross-stack label parity the dataset requires.
//!
//! Exemplar note (verified in Docker, ADR-0003): opentelemetry_sdk 0.32 does NOT
//! populate metric exemplars (the data model has the field but no reservoir/filter
//! pipeline and no OTEL_METRICS_EXEMPLAR_FILTER support). So Payment carries the
//! same exemplar gap as the JS/TS services — app-level RED is real, and the
//! metric↔trace join is served by the bundle's Tempo metrics-generator
//! (traces_spanmetrics_*{service="payment"}). See ARCHITECTURE.md §9.

use anyhow::Result;
use opentelemetry::metrics::{Counter, Histogram};
use opentelemetry_appender_tracing::layer::OpenTelemetryTracingBridge;
use opentelemetry_otlp::{LogExporter, MetricExporter, SpanExporter};
use opentelemetry_sdk::logs::SdkLoggerProvider;
use opentelemetry_sdk::metrics::SdkMeterProvider;
use opentelemetry_sdk::propagation::TraceContextPropagator;
use opentelemetry_sdk::trace::SdkTracerProvider;
use opentelemetry_sdk::Resource;
use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;
use tracing_subscriber::EnvFilter;

/// The instrumentation scope for Payment's own spans, meter, logs and events.
pub const SCOPE: &str = "shoeshop/payment";

/// Holds the providers so they can be flushed/shut down on exit.
pub struct Telemetry {
    tracer: SdkTracerProvider,
    meter: SdkMeterProvider,
    logger: SdkLoggerProvider,
}

impl Drop for Telemetry {
    fn drop(&mut self) {
        let _ = self.tracer.shutdown();
        let _ = self.meter.shutdown();
        let _ = self.logger.shutdown();
    }
}

/// Initialise all four signals. The OTLP endpoint/protocol come from the standard
/// OTEL_EXPORTER_OTLP_* env vars (set in Compose).
pub fn init() -> Result<Telemetry> {
    // service.name from OTEL_SERVICE_NAME, plus service.namespace/deployment.environment
    // from OTEL_RESOURCE_ATTRIBUTES, plus telemetry.sdk.* — and nothing else.
    let resource = Resource::builder().build();

    opentelemetry::global::set_text_map_propagator(TraceContextPropagator::new());

    // Traces.
    let span_exporter = SpanExporter::builder().with_tonic().build()?;
    let tracer = SdkTracerProvider::builder()
        .with_resource(resource.clone())
        .with_batch_exporter(span_exporter)
        .build();
    opentelemetry::global::set_tracer_provider(tracer.clone());

    // Metrics.
    let metric_exporter = MetricExporter::builder().with_tonic().build()?;
    let meter = SdkMeterProvider::builder()
        .with_resource(resource.clone())
        .with_periodic_exporter(metric_exporter)
        .build();
    opentelemetry::global::set_meter_provider(meter.clone());

    // Logs (+ domain events) bridged from `tracing`.
    let log_exporter = LogExporter::builder().with_tonic().build()?;
    let logger = SdkLoggerProvider::builder()
        .with_resource(resource.clone())
        .with_batch_exporter(log_exporter)
        .build();

    // tracing → stdout JSON (so `task logs -- payment` works) + OTLP→Loki bridge.
    let otlp_logs = OpenTelemetryTracingBridge::new(&logger);
    let filter = EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info"));
    tracing_subscriber::registry()
        .with(filter)
        .with(tracing_subscriber::fmt::layer().json())
        .with(otlp_logs)
        .init();

    Ok(Telemetry { tracer, meter, logger })
}

/// App-level RED for the saga workload (per-message processed), keyed by subject +
/// decision result. Cloneable — the OTel instruments are internally ref-counted.
#[derive(Clone)]
pub struct Metrics {
    pub requests: Counter<u64>,
    pub duration: Histogram<f64>,
}

impl Metrics {
    pub fn new() -> Self {
        let meter = opentelemetry::global::meter(SCOPE);
        Metrics {
            requests: meter
                .u64_counter("payment.requests")
                .with_description("Payment commands processed, by subject and result")
                .build(),
            duration: meter
                .f64_histogram("payment.duration")
                .with_unit("s")
                .with_description("Payment processing duration in seconds")
                .build(),
        }
    }
}
